File: src/deepLearning/inference.py
#!/usr/bin/env python3
"""
PyTorch Inference Server for SnipeBT
Loads trained model + scalers, accepts JSON via stdin, returns predictions via stdout
"""

import sys
import os

# Fix Windows encoding for emojis
if sys.platform == 'win32':
    os.environ['PYTHONIOENCODING'] = 'utf-8'
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')
    sys.stderr.reconfigure(encoding='utf-8', errors='replace')

import json
import torch
import pickle
import numpy as np
from pathlib import Path

# Get project root directory (SnipeBT/)
ROOT_DIR = Path(__file__).parent.parent.parent.resolve()

# Import model architecture
from model_pytorch import SnipeBTModel

class InferenceServer:
    def __init__(self, model_path=None, scalers_path=None):
        # Use absolute paths from ROOT_DIR
        if model_path is None:
            model_path = ROOT_DIR / 'best_model_pytorch.pth'
        if scalers_path is None:
            scalers_path = ROOT_DIR / 'scalers.pkl'
        self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        
        # Load model
        print(f"🔥 Loading model from {model_path}...", file=sys.stderr, flush=True)
        self.model = SnipeBTModel().to(self.device)
        checkpoint = torch.load(model_path, map_location=self.device, weights_only=False)
        self.model.load_state_dict(checkpoint['model_state_dict'])
        self.model.eval()
        print(f"✅ Model loaded (Epoch {checkpoint['epoch']}, Val AUC: {checkpoint['val_auc']:.4f})", 
              file=sys.stderr, flush=True)
        
        # Load scalers
        print(f"📊 Loading scalers from {scalers_path}...", file=sys.stderr, flush=True)
        with open(scalers_path, 'rb') as f:
            self.scalers = pickle.load(f)
        print(f"✅ Scalers loaded (candles, context, indicators, patterns)", file=sys.stderr, flush=True)
        
        print("🚀 Inference server ready!", file=sys.stderr, flush=True)
    
    def preprocess(self, raw_input):
        """
        Convert raw input to scaled tensors
        
        Input format:
        {
            "candles": [[open, high, low, close, volume], ...],  // 100 timesteps
            "context": [liquidity, marketCap, holders, age_hours, volume24h],
            "indicators": [rsi, macd, ema_fast, ema_slow, bbands_width],
            "patterns": [has_bullish_pin, has_bearish_pin, has_bullish_engulfing, 
                        has_bearish_engulfing, wick_rejection_ratio, body_to_range_ratio,
                        pattern_confidence, context_score]
        }
        """
        # Convert to numpy arrays
        candles = np.array(raw_input['candles'], dtype=np.float32)  # (100, 5)
        context = np.array(raw_input['context'], dtype=np.float32)  # (5,)
        indicators = np.array(raw_input['indicators'], dtype=np.float32)  # (5,)
        patterns = np.array(raw_input['patterns'], dtype=np.float32)  # (8,)
        
        # Validate shapes
        if candles.shape != (100, 5):
            raise ValueError(f"Expected candles shape (100, 5), got {candles.shape}")
        if context.shape != (5,):
            raise ValueError(f"Expected context shape (5,), got {context.shape}")
        if indicators.shape != (5,):
            raise ValueError(f"Expected indicators shape (5,), got {indicators.shape}")
        if patterns.shape != (8,):
            raise ValueError(f"Expected patterns shape (8,), got {patterns.shape}")
        
        # Scale features
        candles_scaled = self.scalers['candle_scaler'].transform(candles)  # (100, 5)
        context_scaled = self.scalers['context_scaler'].transform(context.reshape(1, -1))[0]  # (5,)
        indicators_scaled = self.scalers['indicator_scaler'].transform(indicators.reshape(1, -1))[0]  # (5,)
        patterns_scaled = patterns  # Patterns already normalized 0-1 or boolean
        
        # Combine context + indicators + patterns
        combined = np.concatenate([context_scaled, indicators_scaled, patterns_scaled])  # (18,)
        
        # Convert to tensors and add batch dimension
        candles_tensor = torch.tensor(candles_scaled, dtype=torch.float32).unsqueeze(0).to(self.device)  # (1, 100, 5)
        combined_tensor = torch.tensor(combined, dtype=torch.float32).unsqueeze(0).to(self.device)  # (1, 18)
        
        return candles_tensor, combined_tensor
    
    def predict(self, raw_input):
        """
        Run inference and return predictions
        
        Returns:
        {
            "profitable": 0.78,        // Probability 0-1
            "max_profit": 4.2,         // Expected max profit %
            "rug_risk": 0.05,          // Probability 0-1
            "confidence": 0.85         // Overall confidence 0-1
        }
        """
        try:
            # Preprocess
            candles, combined = self.preprocess(raw_input)
            
            # Inference
            with torch.no_grad():
                profitable_logit, max_profit_pred, rug_logit = self.model(candles, combined)
                
                # Convert to probabilities
                profitable_prob = torch.sigmoid(profitable_logit).item()
                rug_prob = torch.sigmoid(rug_logit).item()
                max_profit_value = max_profit_pred.item()
                
                # Calculate confidence (distance from 0.5 decision boundary)
                profitable_confidence = abs(profitable_prob - 0.5) * 2  # 0-1 scale
                rug_confidence = abs(rug_prob - 0.5) * 2
                overall_confidence = (profitable_confidence + rug_confidence) / 2
                
                return {
                    "profitable": float(profitable_prob),
                    "max_profit": float(max_profit_value),
                    "rug_risk": float(rug_prob),
                    "confidence": float(overall_confidence)
                }
        
        except Exception as e:
            return {
                "error": str(e),
                "profitable": 0.0,
                "max_profit": 0.0,
                "rug_risk": 1.0,
                "confidence": 0.0
            }
    
    def run(self):
        """
        Main loop: read JSON from stdin, write predictions to stdout
        """
        print("🎯 Listening for inference requests...", file=sys.stderr, flush=True)
        
        for line in sys.stdin:
            line = line.strip()
            if not line:
                continue
            
            try:
                # Parse input
                request = json.loads(line)
                
                # Run prediction
                result = self.predict(request)
                
                # Write result to stdout (TypeScript reads this)
                print(json.dumps(result), flush=True)
            
            except json.JSONDecodeError as e:
                error_result = {
                    "error": f"Invalid JSON: {e}",
                    "profitable": 0.0,
                    "max_profit": 0.0,
                    "rug_risk": 1.0,
                    "confidence": 0.0
                }
                print(json.dumps(error_result), flush=True)
            
            except Exception as e:
                error_result = {
                    "error": f"Inference failed: {e}",
                    "profitable": 0.0,
                    "max_profit": 0.0,
                    "rug_risk": 1.0,
                    "confidence": 0.0
                }
                print(json.dumps(error_result), flush=True)


if __name__ == '__main__':
    server = InferenceServer()
    server.run()

File: src/deepLearning/inference.ts
import * as tf from '@tensorflow/tfjs-node';
import * as fs from 'fs';
import * as path from 'path';

interface CandleData {
    time: number;
    open: number;
    high: number;
    low: number;
    close: number;
    volume: number;
}

interface ContextData {
    liquidity: number;
    marketCap: number;
    holders: number;
    age: number;
    volume24h: number;
}

interface IndicatorData {
    rsi: number;
    macd: number;
    ema_fast: number;
    ema_slow: number;
    bbands_width: number;
}

interface PredictionResult {
    profitable_probability: number;  // 0-1
    max_profit_percent: number;      // Expected max profit %
    rug_risk_probability: number;    // 0-1
    confidence: number;               // Overall confidence 0-1
}

export class DeepLearningInference {
    private model: tf.LayersModel | null = null;
    private scalers: any = null;
    private modelPath: string;
    private scalersPath: string;
    private isReady: boolean = false;
    
    constructor(
        modelPath: string = path.join(__dirname, '../../tfjs_model/model.json'),
        scalersPath: string = path.join(__dirname, '../../scalers.json')
    ) {
        this.modelPath = modelPath;
        this.scalersPath = scalersPath;
    }
    
    /**
     * Load model and scalers
     */
    async initialize(): Promise<void> {
        try {
            console.log('🧠 Loading deep learning model...');
            
            // Load TensorFlow.js model
            this.model = await tf.loadLayersModel(`file://${this.modelPath}`);
            console.log(`✅ Model loaded from ${this.modelPath}`);
            
            // Load scalers (converted from Python pickle to JSON)
            if (fs.existsSync(this.scalersPath)) {
                this.scalers = JSON.parse(fs.readFileSync(this.scalersPath, 'utf-8'));
                console.log(`✅ Scalers loaded from ${this.scalersPath}`);
            } else {
                console.warn(`⚠️ Scalers not found at ${this.scalersPath}, using default scaling`);
                this.scalers = this.getDefaultScalers();
            }
            
            // Warmup prediction
            await this.warmup();
            
            this.isReady = true;
            console.log('✅ Deep learning inference ready!');
            
        } catch (error: any) {
            console.error(`❌ Failed to initialize deep learning: ${error.message}`);
            this.isReady = false;
            throw error;
        }
    }
    
    /**
     * Warmup model with dummy data
     */
    private async warmup(): Promise<void> {
        console.log('🔥 Warming up model...');
        
        const dummyCandles: CandleData[] = Array(100).fill(null).map((_, i) => ({
            time: Date.now() - (100 - i) * 300000,
            open: 1.0,
            high: 1.1,
            low: 0.9,
            close: 1.0,
            volume: 1000
        }));
        
        const dummyContext: ContextData = {
            liquidity: 10000,
            marketCap: 100000,
            holders: 100,
            age: 24,
            volume24h: 50000
        };
        
        const dummyIndicators: IndicatorData = {
            rsi: 50,
            macd: 0,
            ema_fast: 1.0,
            ema_slow: 1.0,
            bbands_width: 0.05
        };
        
        const start = Date.now();
        await this.predict(dummyCandles, dummyContext, dummyIndicators);
        const elapsed = Date.now() - start;
        
        console.log(`✅ Warmup complete (${elapsed}ms)`);
    }
    
    /**
     * Make prediction for a trade setup
     */
    async predict(
        candles: CandleData[],
        context: ContextData,
        indicators: IndicatorData
    ): Promise<PredictionResult | null> {
        if (!this.isReady || !this.model) {
            console.warn('⚠️ Deep learning model not ready');
            return null;
        }
        
        try {
            // Validate inputs
            if (candles.length < 100) {
                console.warn(`⚠️ Not enough candles: ${candles.length}/100`);
                return null;
            }
            
            // Take last 100 candles
            const recentCandles = candles.slice(-100);
            
            // Prepare inputs
            const candlesTensor = this.prepareCandleInput(recentCandles);
            const contextTensor = this.prepareContextInput(context);
            const indicatorsTensor = this.prepareIndicatorInput(indicators);
            
            // Run inference
            const start = Date.now();
            const predictions = this.model.predict([
                candlesTensor,
                contextTensor,
                indicatorsTensor
            ]) as tf.Tensor[];
            
            const elapsed = Date.now() - start;
            
            // Extract predictions
            const [profitableProb, maxProfit, rugRisk] = await Promise.all([
                predictions[0].data(),
                predictions[1].data(),
                predictions[2].data()
            ]);
            
            // Cleanup tensors
            tf.dispose([candlesTensor, contextTensor, indicatorsTensor, ...predictions]);
            
            // Calculate overall confidence
            // High confidence if: profitable likely, good profit expected, low rug risk
            const confidence = this.calculateConfidence(
                profitableProb[0],
                maxProfit[0],
                rugRisk[0]
            );
            
            const result: PredictionResult = {
                profitable_probability: profitableProb[0],
                max_profit_percent: maxProfit[0],
                rug_risk_probability: rugRisk[0],
                confidence
            };
            
            if (process.env.DEBUG) {
                console.log(`🔮 DL Prediction (${elapsed}ms): profit=${(result.profitable_probability * 100).toFixed(1)}%, max=${result.max_profit_percent.toFixed(1)}%, rug=${(result.rug_risk_probability * 100).toFixed(1)}%, conf=${(result.confidence * 100).toFixed(1)}%`);
            }
            
            return result;
            
        } catch (error: any) {
            console.error(`❌ Prediction failed: ${error.message}`);
            return null;
        }
    }
    
    /**
     * Prepare candle data as tensor
     */
    private prepareCandleInput(candles: CandleData[]): tf.Tensor {
        // Extract OHLCV features
        const candleArray = candles.map(c => [
            c.open,
            c.high,
            c.low,
            c.close,
            c.volume
        ]);
        
        // Scale using RobustScaler parameters
        const scaled = this.scalers.candle_scaler 
            ? this.scaleCandles(candleArray, this.scalers.candle_scaler)
            : candleArray;
        
        // Create tensor: shape (1, 100, 5)
        return tf.tensor3d([scaled], [1, 100, 5]);
    }
    
    /**
     * Prepare context data as tensor
     */
    private prepareContextInput(context: ContextData): tf.Tensor {
        const contextArray = [
            context.liquidity,
            context.marketCap,
            context.holders,
            context.age,
            context.volume24h
        ];
        
        // Scale using StandardScaler parameters
        const scaled = this.scalers.context_scaler
            ? this.scaleStandard(contextArray, this.scalers.context_scaler)
            : contextArray;
        
        // Create tensor: shape (1, 5)
        return tf.tensor2d([scaled], [1, 5]);
    }
    
    /**
     * Prepare indicator data as tensor
     */
    private prepareIndicatorInput(indicators: IndicatorData): tf.Tensor {
        const indicatorArray = [
            indicators.rsi,
            indicators.macd,
            indicators.ema_fast,
            indicators.ema_slow,
            indicators.bbands_width
        ];
        
        // Scale using StandardScaler parameters
        const scaled = this.scalers.indicator_scaler
            ? this.scaleStandard(indicatorArray, this.scalers.indicator_scaler)
            : indicatorArray;
        
        // Create tensor: shape (1, 5)
        return tf.tensor2d([scaled], [1, 5]);
    }
    
    /**
     * Scale data using RobustScaler (for candles)
     */
    private scaleCandles(data: number[][], scaler: any): number[][] {
        return data.map(row => 
            row.map((val, i) => {
                const center = scaler.center_[i] || 0;
                const scale = scaler.scale_[i] || 1;
                return (val - center) / scale;
            })
        );
    }
    
    /**
     * Scale data using StandardScaler (for context/indicators)
     */
    private scaleStandard(data: number[], scaler: any): number[] {
        return data.map((val, i) => {
            const mean = scaler.mean_[i] || 0;
            const std = scaler.scale_[i] || 1;
            return (val - mean) / std;
        });
    }
    
    /**
     * Calculate overall confidence from predictions
     */
    private calculateConfidence(
        profitableProb: number,
        maxProfit: number,
        rugRisk: number
    ): number {
        // High confidence = high profitable prob + high expected profit + low rug risk
        
        // Normalize max profit to 0-1 (assume 50% is excellent)
        const profitScore = Math.min(maxProfit / 50, 1);
        
        // Invert rug risk (lower is better)
        const safetyScore = 1 - rugRisk;
        
        // Weighted average
        const confidence = (
            profitableProb * 0.4 +    // 40% weight on profitable
            profitScore * 0.3 +        // 30% weight on expected profit
            safetyScore * 0.3          // 30% weight on safety
        );
        
        return Math.max(0, Math.min(1, confidence));
    }
    
    /**
     * Get default scalers if not available
     */
    private getDefaultScalers(): any {
        return {
            candle_scaler: null,
            context_scaler: null,
            indicator_scaler: null
        };
    }
    
    /**
     * Hot reload model (for auto-refresh after retraining)
     */
    async reload(): Promise<void> {
        console.log('🔄 Reloading deep learning model...');
        
        try {
            // Dispose old model
            if (this.model) {
                this.model.dispose();
            }
            
            // Reload
            await this.initialize();
            
            console.log('✅ Model reloaded successfully!');
            
        } catch (error: any) {
            console.error(`❌ Model reload failed: ${error.message}`);
            throw error;
        }
    }
    
    /**
     * Check if model is ready
     */
    isModelReady(): boolean {
        return this.isReady;
    }
    
    /**
     * Get model info
     */
    getModelInfo(): string {
        if (!this.model) return 'Not loaded';
        
        return `TensorFlow.js model ready (backend: ${tf.getBackend()})`;
    }
}

// Singleton instance
let instance: DeepLearningInference | null = null;

export function getDeepLearningInstance(): DeepLearningInference {
    if (!instance) {
        instance = new DeepLearningInference();
    }
    return instance;
}

export async function initializeDeepLearning(): Promise<void> {
    const dl = getDeepLearningInstance();
    await dl.initialize();
}
